tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -8325,103 +8325,47 @@ Sema::TemplateDeductionResult HLSLExternalSource::DeduceTemplateArgumentsForHLSL
     // TODO: handle template arguments for future intrinsics in a more natural way
 
     // Check Explicit template arguments
+    UINT intrinsicOp = (*cursor)->Op;
     LPCSTR intrinsicName = (*cursor)->pArgs[0].pName;
-    if (ExplicitTemplateArgs) {
-      if (ExplicitTemplateArgs->size() > 1) {
-        getSema()->Diag(ExplicitTemplateArgs->getLAngleLoc(),
-          diag::err_hlsl_unsupported_template_for_intrinsic) << intrinsicName;
-        return Sema::TemplateDeductionResult::TDK_Invalid;
-      }
-      else if (ExplicitTemplateArgs->size() == 1) {
-        if (getSema()->getLangOpts().HLSLVersion < 2018) {
-          getSema()->Diag((*ExplicitTemplateArgs)[0].getLocation(),
-            diag::err_hlsl_unsupported_template_for_intrinsic) << intrinsicName;
-          return Sema::TemplateDeductionResult::TDK_Invalid;
+    bool Is2018 = getSema()->getLangOpts().HLSLVersion >= 2018;
+    bool IsBAB =
+        objectName == g_ArBasicTypeNames[AR_OBJECT_BYTEADDRESS_BUFFER] ||
+        objectName == g_ArBasicTypeNames[AR_OBJECT_RWBYTEADDRESS_BUFFER];
+    bool IsBABLoad = IsBAB && intrinsicOp == (UINT)IntrinsicOp::MOP_Load;
+    bool IsBABStore = IsBAB && intrinsicOp == (UINT)IntrinsicOp::MOP_Store;
+    if (ExplicitTemplateArgs && ExplicitTemplateArgs->size() > 0) {
+      bool isLegalTemplate = false;
+      SourceLocation Loc = ExplicitTemplateArgs->getLAngleLoc();
+      const char *TemplateDiagText = (Is2018 && IsBABLoad) ?
+        "One scalar or vector type argument up to 16 bytes in size" :
+        "No template arguments";
```

Comment:
I realized late, I probably should have changed my advice on the error message after I posted the comment with the code.  It's not best practice to have wording defined in the source here, better to create multiple error messages in DiagnosticSemaKinds.td instead.  Sorry about that.